File: hashmaps/ransom_note.cpp
#include <unordered_map>
#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        if (ransomNote.length() > magazine.length()) return false;
        int arrayMagazine[26] = {0};
        for(int i=0;i<magazine.length();i++){
            arrayMagazine[magazine[i]-'a']++;
        }
        for(int x=0;x<ransomNote.length();x++){
            if(arrayMagazine[ransomNote[x]-'a'] == 0) return false;
            arrayMagazine[ransomNote[x]-'a']--;
        }
        return true;
    }
};
int main() {
    Solution sol;
    cout << sol.canConstruct("a", "b") << endl;
    cout << sol.canConstruct("aa", "ab") << endl;
    cout << sol.canConstruct("aa", "aab") << endl;
    return 0;
}

File: hashmaps/ransom_note.py
from collections import Counter, defaultdict


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        magazine_hashmap = defaultdict(int)
        for x in magazine:
            magazine_hashmap[x]+=1
        
        for y in ransomNote:
            if magazine_hashmap[y]>0:
                magazine_hashmap[y]-=1
            else:
                return False
            
        return True
    
    def canConstruct2(self, ransomNote, magazine):
        #from solutions
            st1, st2 = Counter(ransomNote), Counter(magazine)
            return st1 & st2 == st1
        
    def canConstruct3(self, ransomNote: str, magazine: str) -> bool:
        a = set(ransomNote) 
        for i in a:
            if magazine.count(i)<ransomNote.count(i):
              return False
        return True
solution = Solution()
print(solution.canConstruct3("a","b"))
print(solution.canConstruct3("aa","ab"))
print(solution.canConstruct3("aa","aab"))